### ISSUE TITLE
Support type coercion within array literals

### ISSUE THREAD

**Issue by ben-albrecht** [chapel-lang/chapel#13591]
It can be cumbersome to write out array literals for types like `real` and `complex` or non-default widths:

```chpl
// This doesn't work as of Chapel 1.19
var A = [2.4, 1, 4, 5.2, 11];
var B = [2+1i, 3, 4+2i];
var C = [1:int(8), 2, 5, 7];
```

```chpl
// This is how we have to write it as of Chapel 1.19
var A = [2.4, 1., 4., 5.2, 11.];
var B = [2+1i, 3+0i, 4+2i];
var C = [1:int(8), 2:int(8), 5:int(8), 7:int(8)];
// or
var C = [1, 2, 5, 7]: int(8);
```

One proposal is to allow mixed-type array literals, where the element type is determined by the first value of the literal, and all other elements are coerced to that type. 

With this proposed change, all of the examples from the first code block would work as expected, since the initial element in each array is the desired element type. 

There may be some pitfalls to this approach, but none come to mind at the time of writing. Other proposals are welcome as well.

**Comment by ben-albrecht** [comment]
@wongalvis14 recently encountered this issue, and @bradcray mentioned that we intend to improve it, so I thought I'd open an issue to document it and start us off with an initial design proposal to address it.

**Comment by bradcray** [comment]
My thinking here has traditionally been that we should unify the types of an array literal similar to how we unify the types of returned expressions in a function.  I.e., given an array literal `[e1, ,e2, e3]` the element type of the array literal should be the same as the type of the following function would be:

```chapel
proc foo() {
  if ... then
    return e1;
  if ... then
    return e2;
  if ... then
    return e3;
}
```

(where if a unifying type could not be found, it would be a compiler error, as in the function return case).

Note that this would make your `A` and `B` work as hoped for, but not `C` (I believe) since I think int(64) would be considered to be the unifying type.

**Comment by ben-albrecht** [comment]
> My thinking here has traditionally been that we should unify the types of an array literal similar to how we unify the types of returned expressions in a function.

I like this approach. It's nice that a user doesn't need to be aware of the rules for it to work as expected (in most cases).

> Note that this would make your A and B work as hoped for, but not C (I believe) since I think int(64) would be considered to be the unifying type.

I think I'm OK with that. I think the alternative definition for C that involves a cast is just as syntactically clean.

**Comment by bradcray** [comment]
> I think the alternative definition for C that involves a cast is just as syntactically clean.

I agree and almost said so in my previous post but didn't want to try and railroad you in any specific direction.

**Comment by e-kayrakli** [comment]
My initial reaction when we were discussing with @wongalvis14 was that this should not be supported. Because:

- Things may get a bit too implicit: what would be the behavior if we are talking about objects from the same class hierarchy?, for example. It may make the code easier to write but harder to read.
- This might create the impression (at least on a beginner) that Chapel's type system is similar to Python's. i.e. you can mix types in an array (list for python)
- Potential impact on compilation time (probably not as big of a concern as the others)

None of these are huge concerns and I wouldn't object to make this change.

**Comment by ben-albrecht** [comment]
@npadmana and I were recently wishing for this feature during development of tests in  https://github.com/chapel-lang/chapel/pull/15626

Regarding @e-kayrakli's comment, the only one that concerns me at the moment is:

> Things may get a bit too implicit: what would be the behavior if we are talking about objects from the same class hierarchy?, for example. It may make the code easier to write but harder to read.

We might want to implement this behavior for primitive types only, or only exclude class types to avoid this issue.

Always relying on the first element type is another solution to avoiding confusion here.

**Comment by bradcray** [comment]
I still stand by the approach in my [previous comment ](https://github.com/chapel-lang/chapel/issues/13591#issuecomment-517000027) and don't think we should shy away from it or special-case it for non-classes.  It wouldn't make sense to me to say "we can unify these class types in the case of [inferring a function return type](https://tio.run/##dY5BCoMwFET3nmKWyabQdqd0pfEeEn40EJKQpIiUnj1NK6UodvV5M4/hy2nwZHKWzio9opyYoMww4oZzU1XSDDGixaN6fqGrtyx23NcQK/vgJJRzjBcGtAJbp8s2R5rIlhQIlO7BwtKMlvGmZGQibfTLsd790a/Huvjp@6p/V@XlOehExrLP16e0eKpjCtqOvMn5BQ), but not in the context of an array value."  Either case has similar potential challenges to readability, but the code author can help if they choose to by declaring the static type of the thing in question.

More generally, I think all kinds of language features have the potential for leading to abuse or confusion if pushed to extremes, but I think the rule is not that difficult while also being similar to what we do (or should do) in other multi-expression inference cases, and is unlikely to result in the confusing / broken sorts of behavior in common usage.

**Comment by e-kayrakli** [comment]
Re-reading this discussion, my position has not changed that much. Things may get a bit implicit if we do support this. But I am not against it or anything. And, Brad's point about function type inference make sense to me.
